### PR TITLE
Improve WriteScope in ir tree write phase

### PR DESCRIPTION
**Pull request by jdconrad**
This change moves loose ASM labels into the WriteScope as opposed to having these as member data on the nodes when they don't make sense for all executed phases. This adds the required scopes such as loop scope and try scope to the WriteScope to account for this.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java [hunk 1]
```diff
@@ -96,15 +92,14 @@ protected void write(ClassWriter classWriter, MethodWriter methodWriter, WriteSc
         methodWriter.visitVarInsn(variable.getAsmType().getOpcode(Opcodes.ISTORE), variable.getSlot());
 
         if (blockNode != null) {
-            blockNode.continueLabel = continueLabel;
-            blockNode.breakLabel = breakLabel;
-            blockNode.write(classWriter, methodWriter, writeScope);
+            blockNode.write(writeScope.newTryScope(null, null, null));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java:95]: This is weird.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java [hunk 2]
```diff
@@ -96,15 +92,14 @@ protected void write(ClassWriter classWriter, MethodWriter methodWriter, WriteSc
...
         }
 
-        methodWriter.visitTryCatchBlock(begin, end, jump, variable.getAsmType().getInternalName());
+        methodWriter.visitTryCatchBlock(
+                writeScope.getTryBeginLabel(), writeScope.getTryEndLabel(), jump, variable.getAsmType().getInternalName());
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java:99]: Why does method stuff things into the write scope to just pull them out?  Maybe `visitTryCatchBlock` can take in a `WriteScope`?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java [hunk 3]
```diff
@@ -19,15 +19,85 @@
...
-    /* ---- begin node data ---- */
+    /* ---- begin decorations ---- */
+
+    public interface IRDecoration {
+
+    }
+
+    private final Map<Class<? extends IRDecoration>, IRDecoration> decorations = new HashMap<>();
+
+    @SuppressWarnings("unchecked")
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:41]: This doesn't seem relevant to this PR.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 4]
```diff
@@ -1727,6 +1727,7 @@ public void visitCall(ECall userCallNode, ScriptScope scriptScope) {
             }
 
             InvokeCallNode irInvokeCallNode = new InvokeCallNode(userCallNode.getLocation());
+
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1730]: Why?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/symbol/WriteScope.java [hunk 5]
```diff
@@ -64,21 +66,138 @@ public int getSlot() {
...
+        this.breakLabel = parent.breakLabel;
+        this.tryBeginLabel = parent.tryBeginLabel;
+        this.tryEndLabel = parent.tryEndLabel;
+        this.catchesEndLabel = parent.catchesEndLabel;
+        this.nextSlot = parent.nextSlot;
     }
 
-    public WriteScope newScope() {
-        return new WriteScope(this, nextSlot);
+    protected WriteScope(WriteScope parent, MethodWriter methodWriter) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/symbol/WriteScope.java:103]: jdoc for each of these protected constructors, just a line.

**Review by stu-elastic — APPROVED** [review]
Address comments, then g2g

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java:99]: This is the actual method to write the ASM instruction for a try/catch block and requires all these labels.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CatchNode.java:95]: Agreed, updated scoping to improve the handling for this unique case.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:41]: Yup, this is for the next PR. (Oops!) Removed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1730]: Removed. Accidental.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/symbol/WriteScope.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/symbol/WriteScope.java:103]: Done.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review!
